Create tests for bugs reported about empty lines in strings vs indentation (#159)

lelutin opened the pull request:
Those are all of the examples that match the reported indentation errors.

The very strange thing is that all of the new tests fail, but they fail differently in vader than how they do in an interactive (n)vim session. The result that we get in those tests is usually that the string contents get wrongly moved to fit indentation. However, when I run the indentation command in the editor, the string contents don't change but things afterwards get indented at the wrong level. I'm not sure why the tests don't behave in the same way.

lelutin committed:
Create tests for bugs reported about empty lines in strings vs indentation (#159)

Those are all of the examples that match the reported indentation
errors.

The very strange thing is that all of the new tests fail, but they fail
differently in vader than how they do in an interactive (n)vim session.
The result that we get in those tests is usually that the string
contents get wrongly moved to fit indentation. However, when I run the
indentation command in the editor, the string contents don't change but
things afterwards get indented at the wrong level. I'm not sure why the
tests don't behave in the same way.

lelutin commented:
@alexjfisher fyi I've created new tests with the reproducing examples you reported in #159 

like I described in the commit message vader.vim is failing on all of these, but not in the same way that we've both seen :(  I wonder if I got something wrong. but if I run indentation with `=` on the test files I still see the same behavior that we both saw, not what the tests report as their result.

alexjfisher commented:
For the life of me, I can't work out what's going on in the tests, and why it's different.

alexjfisher commented:
Now I have even less of a clue as to what's going on.
Why would the following change the behaviour of the new tests??

```diff
--- a/indent/puppet.vim
+++ b/indent/puppet.vim
@@ -47,6 +47,7 @@ function! s:OpenBrace(lnum)
 endfunction

 function! s:InsideMultilineString(lnum)
+    let g:wtf = synID(line("$"), 1, 0)
     return synIDattr(synID(a:lnum, 1, 0), 'name') =~? 'string'
 endfunction
 ```
 
 With this simple test:
 ```
 Given puppet (indented heredoc):
  class blah {
    $value = @(HERE)
      Simple indented heredoc
      | HERE
  }

Do (full text indent with '='):
  gg=G

Expect puppet (should remain the same):
  class blah {
    $value = @(HERE)
      Simple indented heredoc
      | HERE
  }
  ```
  
  without my seemingly pointless change...
  
  ```
  Starting Vader: 1 suite(s), 1 case(s)
  Starting Vader: /home/alex/test-vim-puppet/vim-puppet/test/indent/indented_heredoc.vader
    (1/1) [  GIVEN] indented heredoc
    (1/1) [     DO] full text indent with '='
    (1/1) [ EXPECT] (X) should remain the same
      - Expected:
          class blah {
            $value = @(HERE)
              Simple indented heredoc
              | HERE
          }
      - Got:
          class blah {
            $value = @(HERE)
            Simple indented heredoc
            | HERE
          }
  Success/Total: 0/1
Success/Total: 0/1 (assertions: 0/0)
```

But with the dumb patch
```
Starting Vader: 1 suite(s), 1 case(s)
  Starting Vader: /home/alex/test-vim-puppet/vim-puppet/test/indent/indented_heredoc.vader
    (1/1) [  GIVEN] indented heredoc
    (1/1) [     DO] full text indent with '='
    (1/1) [ EXPECT] should remain the same
  Success/Total: 1/1
Success/Total: 1/1 (assertions: 0/0)
```

lelutin commented:
wat.. I'm thouroughly confused as to what the unit test plugin is doing here :(

it's too bad since it was working fine for most other tests. also from what I've seen so far, it's pretty much the only one that lets us test what (n)vim is doing when running actual commands on a test buffer.

I've also tried to load nvim with the test vimrc file with `nvim -u test/init.vim` and doing the indentation manually with `gg=G` on a copy of one of our new tests and I'm still getting the behaviour we've both observed, not the one that's displayed by vader.vim.

we may have stumbled upon a vader.vim bug there

alexjfisher commented:
I think it's just a race condition.  The syntax hasn't been evaluated by the time `gg=G` is run.

alexjfisher commented on test/indent/comments_strings.vader:
```diff
@@ -63,3 +63,99 @@ Do (New line):
+  }
+
+Do (full text indent with '='):
+  gg=G
```
The `synID` function might return the wrong values if syntax evaluation hasn't happened in time.  Either sleep, or better yet, do the following?
```suggestion
  :syntax sync fromstart
  gg=G
```

lelutin committed:
Update test/indent/comments_strings.vader

Try and wait for syntax to be fully evaluated before running the indexing command

Co-authored-by: Alexander Fisher <alex@linfratech.co.uk>

lelutin committed:
Implement the same fix as the previous commit for all new tests

Things really did change with the addition of the wait for the syntax.
Apparently as @alexjfisher said, the format function needs its calls to
get the correct syntax ID so we need the syntax file to be fully loaded
before trying to format things.

lelutin commented:
@alexjfisher nice that seems to help, thanks! and your reasoning for the need to call that sync to wait for syntax is sound. it's been a long while since I touched the code in this module.. I've implemented the same thing on all tests.

now the new mystery is ... well all of the tests are now passing which is not what we were expecting so far:  the tests are not replicating the odd removals of spaces on empty lines and the odd indentation after a string with an empty line  :thinking:

Note: the vim 8 tests that fail seem to be unrelated to this MR.
